pkg/relayer/proxy/rings.go

```diff
@@ -0,0 +1,108 @@
+// TODO(@h5law): Move all this logic out into a shared package
+package proxy
+
+import (
+	"context"
+	"fmt"
+
+	ring_secp256k1 "github.com/athanorlabs/go-dleq/secp256k1"
+	ringtypes "github.com/athanorlabs/go-dleq/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	accounttypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	ring "github.com/noot/ring-go"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+)
+
+// getRingForAppAddress returns the RingSinger used to sign relays. It does so by fetching
+// the latest information from the application module and creating the correct ring.
+// This method also caches the ring's public keys for future use.
+func (rp *relayerProxy) getRingForAppAddress(ctx context.Context, appAddress string) (*ring.Ring, error) {
+	// lock the cache for reading
+	rp.ringCacheMutex.RLock()
+
+	// check if the ring is in the cache
+	points, ok := rp.ringCache[appAddress]
+	rp.ringCacheMutex.RUnlock() // unlock the cache incase not found in cache
+	var err error
+	if !ok {
+		// if the ring is not in the cache, get it from the application module
+		points, err = rp.getDelegatedPubKeysForAddress(ctx, appAddress)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// create the ring from the points
+	return newRingFromPoints(points)
+}
+
+// newRingFromPoints creates a new ring from a slice of points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for the given application
+// address, by querying the portal module for it's delegated pubkeys
+func (rp *relayerProxy) getDelegatedPubKeysForAddress(
```

Comment:
This file seems to be a complete copy-paste of `pkg/appgateserver/rings.go`. What's up with that?

Reply:
See top line, this will be refactored into its on `pkg` subdir as this logic should be encapsulated but I think thats not work for today copy-pasta is better as the caching logic etc will be more complex and require more time to implement